cpp/include/cudf/dictionary/dictionary_factories.hpp

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_view.hpp>
+#include <cudf/null_mask.hpp>
+
+
+namespace cudf
+{
+
+/**
+ * @brief Construct a dictionary column from a column.
+ *
+ * Unique elements are managed in a dictionary.
+ * The output column will have indices with the same count
+ * as the input column.
+ *
+ * ```
+ * c = [429,111,213,111,213,429,213]
+ * d = make_dictionary_column(c)
+ * d now has dictionary [111,213,429] and indices [2,0,1,0,1,2,1]
+ * ```
+ *
+ * @param[in] column The column data to build the dictionary from.
+ * @param[in] mr Optional resource to use for device memory allocation.
+ * @param[in] stream Optional stream on which to issue all memory allocation and
+ * device kernels.
+ */
+std::unique_ptr<column> make_dictionary_column(
+    column_view const& column,
+    rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource(),
+    cudaStream_t stream = 0);
+
+/**
+ * @brief Construct a dictionary column by using the provided keys
+ * and indices.
+ *
+ * The keys_column must contain no nulls. The keys should be
+ * unique and sorted in ascending order.
```

Comment:
We should relax this constraint for Python's use cases. @kkraus14 may be able to help me out on how to best word this, but something like:
```suggestion
 * unique and it is assumed the elements in `keys` are in a strict, total order. Meaning, 
 * `keys[i]` is _ordered before_ `keys[i+1]` for all `i in [0,n-1)` where `n` is the number of keys.
```

It's been a while since set theory, but I think I got my terms correct. Would be helpful if someone could double check. 

This allows for a user to specify a custom ordering that may not necessarily be a traditional sorted order.

Reply:
I would reword as "it is assumed the elements in `keys` are in a logical order regardless of their lexical order"

Reply:
> I would reword as "it is assumed the elements in `keys` are in a logical order regardless of their lexical order"

This only works if you do not ever add, remove, or otherwise combine with new keys once the dictionary column is created. The sort order is used to find matching/duplicate keys when combining with another set of keys either for updating or removing.

Reply:
I think I like my wording better. "logical order" doesn't tell you anything about what kind of order,  and "lexical order" is really only meaningful for strings.